### PR TITLE
adjust documented inline-asm register constraints

### PR DESCRIPTION
**Pull request by Lokathor**
This change more clearly specifies how `reg` and `reg_thumb` work with ARM, Thumb2, and Thumb1 code.

Based upon the [llvm documentation](https://llvm.org/docs/LangRef.html#supported-constraint-code-list) for register constraint codes.
To be clear, this just updates the docs to match what already happens with rustc/llvm.
No change in the compiler is required to make it match this new documentation.

### CODE REVIEW

**Comment by Lokathor** [comment]
(as discussed at the 2021-11-09 T-lang meeting)

cc @Amanieu and @joshtriplett

**Comment by Lokathor** [comment]
It's my understanding that Amaneau would prefer that the `reg` class not do separate things on thumb1 as compared to thumb2 and ARM, but LLVM already operates this way. If the full range of registers is allowed on thumb1 then it just leads to strange compile errors. Registers above r0-r7 aren't allowed in thumb1 with instructions other than `add`, `mov`, and `cmp`, so those registers shouldn't be allocated to thumb1 inline assembly by default.

If we want to allow the high registers to be used with thumb1 inline assembly, a separate register class could be made for that, but note that LLVM doesn't even actually support that use case at this time. There is not any register class you can provide to LLVM to make give you the high registers (r8 and above) with thumb1 code.

File: src/doc/unstable-book/src/library-features/asm.md [hunk 1]
```diff
@@ -562,9 +562,12 @@ Here is the list of currently supported register classes:
 | AArch64 | `vreg` | `v[0-31]` | `w` |
 | AArch64 | `vreg_low16` | `v[0-15]` | `x` |
 | AArch64 | `preg` | `p[0-15]`, `ffr` | Only clobbers |
-| ARM | `reg` | `r[0-12]`, `r14` | `r` |
-| ARM (Thumb) | `reg_thumb` | `r[0-r7]` | `l` |
+| ARM (ARM) | `reg` | `r[0-12]`, `r14` | `r` |
+| ARM (Thumb2) | `reg` | `r[0-12]`, `r14` | `r` |
+| ARM (Thumb1) | `reg` | `r[0-r7]` | `r` |
```
Comment (joshtriplett) [inline src/doc/unstable-book/src/library-features/asm.md:567]:
```suggestion
| ARM (Thumb1) | `reg` | `r[0-7]` | `r` |
```

File: src/doc/unstable-book/src/library-features/asm.md [hunk 2]
```diff
@@ -562,9 +562,12 @@ Here is the list of currently supported register classes:
...
 | ARM (ARM) | `reg_thumb` | `r[0-r12]`, `r14` | `l` |
+| ARM (Thumb2) | `reg_thumb` | `r[0-r7]` | `l` |
+| ARM (Thumb1) | `reg_thumb` | `r[0-r7]` | `l` |
```
Comment (joshtriplett) [inline src/doc/unstable-book/src/library-features/asm.md:570]:
```suggestion
| ARM (Thumb2) | `reg_thumb` | `r[0-7]` | `l` |
| ARM (Thumb1) | `reg_thumb` | `r[0-7]` | `l` |
```

**Review by joshtriplett — CHANGES_REQUESTED** [review]

**Comment by Amanieu** [comment]
I'm actually starting to have second thoughts on the value of having a separate `reg_thumb` register class. I'm probably going to remove it before stabilization.

In any case, this PR accurately describes what currently happens.

@bors r+ rollup

**Pull request merged**